File: ABB-setup/dv-processing-rel_1.7/docs/source/assets/code_samples/python/event_stream_slicing/sample1.py
import dv_processing as dv
from datetime import timedelta

# Initialize slicer, it will have no jobs at this time
slicer = dv.EventStreamSlicer()


def print_time_interval(events: dv.EventStore):
    # Print the time duration received by this method
    print(f"* Received event with {events.duration()} duration in time-based slicing")


# Register this method to be called every 33 millisecond worth of event data
time_job_id = slicer.doEveryTimeInterval(timedelta(milliseconds=33), print_time_interval)


def print_event_number(events: dv.EventStore):
    # Print the number of events received here
    print(f"# Received {events.size()} events in number-based slicing")


# Register this method to be called every 100 events
number_job_id = slicer.doEveryNumberOfEvents(100, print_event_number)

# Implement data generation; The following loop will generate 10 stores
# of events, each containing 100 events within 10 millisecond duration.
for i in range(10):
    # Generate 100 events within 20 millisecond interval. These will be sliced correctly by the slicer.
    store = dv.data.generate.uniformEventsWithinTimeRange(i * 20000, timedelta(milliseconds=20), (100, 100), 100)

    # Now push the store into the slicer, the data contents within the store
    # can be arbitrary, the slicer implementation takes care of correct slicing
    # algorithm and calls the previously registered callbacks accordingly.
    slicer.accept(store)

    # When a packet with index 5 is reached, modify the parameters
    if i == 5:
        # Modify time range to 10 milliseconds instead of 33
        slicer.modifyTimeInterval(time_job_id, timedelta(milliseconds=10))
        # Modify number to 200 instead of 100
        slicer.modifyNumberInterval(number_job_id, 200)

File: ABB-setup/dv-processing-rel_1.7/docs/source/assets/code_samples/c++/event_stream_slicing/sample1.cpp
#include <dv-processing/core/core.hpp>
#include <dv-processing/data/generate.hpp>

int main() {
	// Use this namespace to enable literal time expression from the chrono library
	using namespace std::chrono_literals;

	// Initialize slicer, it will have no jobs at this time
	dv::EventStreamSlicer slicer;

	// Register this method to be called every 33 millisecond worth of event data
	int timeJobId = slicer.doEveryTimeInterval(33ms, [](const dv::EventStore &events) {
		std::cout << "* Received events time-based slicing: " << events << std::endl;
	});

	// Register this method to be called every 100 events
	int numberJobId = slicer.doEveryNumberOfElements(100, [](const dv::EventStore &events) {
		std::cout << "# Received events in number-based slicing: " << events << std::endl;
	});

	// Implement data generation; The following loop will generate 10 packets of events,
	// each containing 100 events within 20 millisecond duration.
	for (int i = 0; i < 10; i++) {
		// Generate 100 events within 20 millisecond interval. These will be sliced correctly by the slicer.
		const auto store = dv::data::generate::uniformEventsWithinTimeRange(i * 20'000, 20ms, cv::Size(100, 100), 100);

		// Now push the store into the slicer, the data contents within the store
		// can be arbitrary, the slicer implementation takes care of correct slicing
		// algorithm and calls the previously registered callbacks accordingly.
		slicer.accept(store);

		// When a packet with index 5 is reached, modify the parameters
		if (i == 5) {
			// Modify time range to 10 milliseconds instead of 33
			slicer.modifyTimeInterval(timeJobId, 10ms);
			// Modify number to 200 instead of 100
			slicer.modifyNumberInterval(numberJobId, 200);
		}
	}
}
